### PR TITLE
[REF] mail, *: refactor Discuss

### PR DESCRIPTION
**Pull request by seb-odoo**
This commit refactors discuss code to use OWL components
and new tools and services in web/.
Functionally, Discuss app should work relatively the same as
before this commit.

Co-authored-by: Alexandre (aku) <aku@odoo.com>
Co-authored-by: Didier (did) <did@odoo.com>
Co-authored-by: Géry (ged) <ged@odoo.com>
Co-authored-by: Louis (wil) <wil@odoo.com>
Co-authored-by: Maël (mapa) <mapa@odoo.com>
Co-authored-by: Maryam (maki) <maki@odoo.com>
Co-authored-by: Sébastien (seb) <seb@odoo.com>
Co-authored-by: Thanh (tso) <tso@odoo.com>
Co-authored-by: Matthieu (tsm) <tsm@odoo.com>
Co-authored-by: Zelong (zel) <zel@odoo.com>

closes https://github.com/odoo/odoo/pull/110188

Related:
https://github.com/odoo/enterprise/pull/38058
https://github.com/odoo/upgrade/pull/4423

### CODE REVIEW

**Comment by seb-odoo** [comment]
@robodoo override=ci/security

enterprise/documents/static/src/views/hooks.js:496
markup(
=> same LOC (just prettify), dynamic content is escaped in place

enterprise/voip/static/src/js/registerer.js:87
this.voip.triggerError(markup${errorMessage}</br></br>${help});
=> error message content has some parts (`reasonPhrase`) coming from sip.js lib, which looks trustworthy

**Comment by seb-odoo** [comment]
@robodoo override=ci/security

`odoo/addons/mail/static/src/core/message_service.js:41`
`message.body = markup(data.body);`
=> message data is user-generated and come from server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:138`
`body: messageData.body ? markup(messageData.body) : messageData.body,`
=> message data is user-generated and come from server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:255`
`Object.assign(notif.payload, { body: markup(notif.payload.body) })`
=> message transient data come immediately from server, caller is responsible of sending safe content

`odoo/addons/mail/static/src/core/messaging_service.js:267`
`const data = Object.assign(notif.payload, { body: markup(notif.payload.body) });`
=> message data is user-generated and come server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:396`
`body: markup(message.body),`
=> message data is user-generated and come from server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:449`
`messageData.parentMessage.body = markup(messageData.parentMessage.body);`
=> message data is user-generated and come from server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:452`
`body: markup(messageData.body),`
=> message data is user-generated and come from server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:535`
`body: messageData.body ? markup(messageData.body) : messageData.body,`
=> record/insert notification of "Message" are user-generated message content and come from server, which have been sanitized

`odoo/addons/mail/static/src/core/messaging_service.js:598`
`body: markup(preview.last_message.body),`
=> message data in preview is user-generated and sanitized (come from server)

`odoo/addons/mail/static/src/core/thread_service.js:181`
`? markup(data.parentMessage.body)`
=> message data fetched from server, is user-generated and sanitized (come from server)

`odoo/addons/mail/static/src/core/thread_service.js:185`
`Object.assign(data, { body: data.body ? markup(data.body) : data.body }),`
=> message data fetched from server, is user-generated and sanitized (come from server)

`odoo/addons/mail/static/src/core/thread_service.js:716`
`body: markup(await prettifyMessageContent(body, validMentions)),`
=> body is self-generated, not broadcast to other users. Valid mentions might contain
risky values (partner names), but it's escaped and safely transformed in `prettifyMessageContent`
implementation (this function is basically a move).

`odoo/addons/mail/static/src/core/thread_service.js:724`
`? markup(data.parentMessage.body)`
=> message data fetched from server, is user-generated and sanitized (come from server)

`odoo/addons/mail/static/src/core/thread_service.js:728`
`Object.assign(data, { body: markup(data.body) })`
=> message data fetched from server, is user-generated and sanitized (come from server)

`odoo/addons/mail/static/src/utils/format.js:47`
`children = $("<div>").html(string).contents();`
=> code move, `<` are escaped before processing

`odoo/addons/mail/static/src/utils/format.js:50`
`.html("<pre>" + string + "</pre>")`
=> code move, `<` are escaped before processing

`odoo/addons/mail/static/src/utils/format.js:110`
`div.innerHTML = linkified;`
=> code move, escaping done inside linkify function

`odoo/addons/mail/static/src/utils/format.js:211`
`div.innerHTML = htmlString;`
=> code move, inner processing of function inside a fragment, not actually inserted into dom (textContent extracted and later printed escaped)

`odoo/addons/mail/static/src/utils/format.js:213`
`div.innerHTML = <pre>${htmlString}</pre>;`
=> code move, inner processing of function inside a fragment, not actually inserted into dom (textContent extracted and later printed escaped)

`odoo/addons/mail/static/src/web/thread_service_patch.js:56`
`activity.note = markup(activity.note);`
=> activity note data is user-generated and sanitized from server

`odoo/addons/mail/static/tests/helpers/test_utils.js:316`
`dropdownDiv.innerHTML =`
=> static template

`odoo/addons/mail/static/tests/mail_utils_tests.js:72`
`div.innerHTML = linkified;`
=> code move, in test so irrelevant

`odoo/addons/mail/static/tests/mail_utils_tests.js:87`
`div.innerHTML = linkified;`
=> code move, in test so irrelevant

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

File: addons/mail/models/res_partner.py [hunk 1]
```diff
@@ -334,19 +334,22 @@ def get_mention_suggestions(self, search, limit=8, channel_id=None):
...
+        if excluded_ids is None:
+            excluded_ids = []
         users = self.env['res.users'].search([
             ('id', '!=', self.env.user.id),
             ('name', 'ilike', name),
             ('active', '=', True),
             ('share', '=', False),
         ], order='name, id', limit=limit)
-        return list(users.partner_id.mail_partner_format().values())
+        return list(users.filtered(lambda user: user.partner_id.id not in excluded_ids).partner_id.mail_partner_format().values())
```
Comment (KangOl) [inline addons/mail/models/res_partner.py:355]: Put this condition in the search domain

Comment (alexkuhn) [reply addons/mail/models/res_partner.py]: done

File: addons/mail/models/res_partner.py [hunk 2]
```diff
@@ -334,19 +334,23 @@ def get_mention_suggestions(self, search, limit=8, channel_id=None):
...
         # extension via an invite. As both of these actions require the 'create' access
         # right, we check this specific ACL.
+        if excluded_ids is None:
+            excluded_ids = []
         users = self.env['res.users'].search([
             ('id', '!=', self.env.user.id),
             ('name', 'ilike', name),
             ('active', '=', True),
             ('share', '=', False),
+            ('partner_id.id', 'not in', excluded_ids)
```
Comment (KangOl) [inline addons/mail/models/res_partner.py:354]:
```suggestion
            ('partner_id', 'not in', excluded_ids)
```

Comment (alexkuhn) [reply addons/mail/models/res_partner.py]: done

**Comment by alexkuhn** [comment]
@robodoo r+

**Pull request closed without merging**